File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque


def solution(maps):

    start, lever, exit = (0, 0), (0, 0), (0, 0)

    for i in range(len(maps)):
        for j in range(len(maps[0])):
            if maps[i][j] == 'S':
                start = (i, j)
            elif maps[i][j] == 'L':
                lever = (i, j)
            elif maps[i][j] == 'E':
                exit = (i, j)

    start_to_level = bfs(maps, start, lever)
    lever_to_exit = bfs(maps, lever, exit)

    if start_to_level == -1 or lever_to_exit == -1:
        return -1

    return start_to_level + lever_to_exit

def bfs(maps, start, end):
    n, m = len(maps), len(maps[0])

    sx, sy = start
    ex, ey = end

    visit = [[False] * m for _ in range(n)]
    visit[sx][sy] = True

    qu = deque([(sx, sy, 0)])

    dx = (-1, 1, 0, 0)
    dy = (0, 0, -1, 1)

    while qu:
        x, y, count = qu.popleft()

        if x == ex and y == ey:
            return count

        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]

            if nx < 0 or ny < 0 or nx >= n or ny >= m:
                continue
            if visit[nx][ny] or maps[nx][ny] == 'X':
                continue

            visit[nx][ny] = True
            qu.append((nx, ny, count + 1))
    
    return -1

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    public int solution(String[] maps) {
        int startX = 0, startY = 0;
        int leverX = 0, leverY = 0;
        int exitX = 0, exitY = 0;

        for (int i = 0; i < maps.length; i++) {
            for (int j = 0; j < maps[i].length(); j++) {
                char c = maps[i].charAt(j);
                if (c == 'S') {
                    startX = i;
                    startY = j;
                } else if (c == 'L') {
                    leverX = i;
                    leverY = j;
                } else if (c == 'E') {
                    exitX = i;
                    exitY = j;
                }
            }
        }

        int startToLever = bfs(maps, startX, startY, leverX, leverY);
        int leverToExit = bfs(maps, leverX, leverY, exitX, exitY);

        if (startToLever == -1 || leverToExit == -1) {
            return -1;
        }

        return startToLever + leverToExit;
    }
    
    private int bfs(String[] maps, int sx, int sy, int ex, int ey) {

        int n = maps.length;
        int m = maps[0].length();

        boolean[][] visit = new boolean[n][m];
        visit[sx][sy] = true;

        Queue<Integer> qu = new ArrayDeque<>();
        qu.offer(sx);
        qu.offer(sy);
        qu.offer(0);

        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        while (!qu.isEmpty()) {
            int x = qu.poll();
            int y = qu.poll();
            int count = qu.poll();

            if (x == ex && y == ey) {
                return count;
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx >= n || ny >= m) {
                    continue;
                }
                if (visit[nx][ny] || maps[nx].charAt(ny) == 'X') {
                    continue;
                }
                visit[nx][ny] = true;
                qu.offer(nx);
                qu.offer(ny);
                qu.offer(count + 1);
            }
        }

        return -1;
    }
}